feat: Add option to show urgency. More space before pencil icon.

Cito opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

# Description

This PR adds an option that allows showing the urgency of each task (#1111).

It also extends the existing syntax in a backward compatible way: In addition to the command `hide`, you can now also use the command `show`. The urgency is hidden by default, so to show it, you need to use the command `show urgency`. If you use `show` with an options that is shown by default, or `hide` with an option that is hidden by default, this will ne a no-op.

This PR also slightly simplifies the HTML elements that are postponed to the task. Before, the edit button was inside the span for the backlink, which does not make much sense. The elements for the backlink, the edit button and the urgency score are now simply appended as childs of the existing list item, not nested into each other.

I also removed an unnecessary `trim()` and `toLowerCase()` when evaluating the options, since the used regex guarantees the strings to be lower case and trimmed already.

Let me know if I shall add tests as well and what would be the best way to do it.

Also, mabye the order of the edit button and the urgency score should be swapped, but that would be a very simple change.

## Motivation and Context

See #1111.

## How has this been tested?

I have been using the sample vault to check this out and added and removed `show urgency` to the queries to test.

## Screenshots

![sample-show-urgency](https://user-images.githubusercontent.com/464599/189551946-77fb1155-fbbd-4774-9802-f118bb27b1ce.png)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

Changes visible to users:

- [ ] **Bug fix** (prefix: `fix` - non-breaking change which fixes an issue)
- [X] **New feature** (prefix: `feat` - non-breaking change which adds functionality)
- [ ] **Breaking change** (prefix: `feat!!` or `fix!!` - fix or feature that would cause existing functionality to not work as expected)
- [x] **Documentation** (prefix: `docs` - improvements to any documentation content)
- [ ] **Sample vault** (prefix: `vault` - improvements to the [Tasks-Demo sample vault](https://github.com/obsidian-tasks-group/obsidian-tasks/tree/main/resources/sample_vaults/Tasks-Demo))

Internal changes:

- [X] **Refactor** (prefix: `refactor` - non-breaking change which only improves the design or structure of existing code, and making no changes to its external behaviour)
- [ ] **Tests** (prefix: `test` - additions and improvements to unit tests and the smoke tests)
- [ ] **Infrastructure** (prefix: `chore` - examples include GitHub Actions, issue templates)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [X] My code follows the code style of this project and passes `yarn run lint`.
- [X] My change requires a change to the documentation.
- [X] I have [updated the documentation](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md#updating-documentation) accordingly.
- [ ] My change has adequate [Unit Test coverage](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md#maintaining-the-tests).

## Terms

<!--
By submitting this pull request, you must agree to follow our
[contributing guide](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md) and
[Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md).
Put an x in the boxes to confirm you agree.
-->

- [X] My contribution follow this project's [contributing guide](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CONTRIBUTING.md)

- [X] I agree to follow this project's [Code of Conduct](https://github.com/obsidian-tasks-group/obsidian-tasks/blob/main/CODE_OF_CONDUCT.md)

claremacrae commented:
Hi @Cito this is great - thank you very much indeed.

I won't have time for a full code review yet, but I did try it out in the Tasks-Demo vault, and have a few observations... 

## Presentation

1. Can the default styling be made so that it has more contrast in Light mod:

  ![1139 - feat - show urgency - unreadable in light mode](https://user-images.githubusercontent.com/4840096/189592472-f2a9b28f-b2fd-4525-a249-2c4ce7d492e5.png)

  - I think there is some accessibility score scheme for contrast in colours for readability - it would be good to satisfy that.

2. Can the score for done tasks be just as readable as the score for not-done tasks.

  ![1139 - feat - show urgency - done in light](https://user-images.githubusercontent.com/4840096/189592924-f5b4c747-737b-4f3e-b09a-870f04357e7f.png)

  ![1139 - feat - show urgency - done in dark](https://user-images.githubusercontent.com/4840096/189592947-4661f208-46be-49a1-9462-5103f8201173.png)

3. Please keep the pencil as the right-most thing on the line, and I think as backlinks vary so much in length, it would also be better if the score went before the backlink... so that it will be easier to locate the score with a glance.

Hahah - I see you thought much the same:

> Also, mabye the order of the edit button and the urgency score should be swapped, but that would be a very simple change.

## Docs

 >It also extends the existing syntax in a backward compatible way: In addition to the command `hide`, you can now also use the command `show`. The urgency is hidden by default, so to show it, you need to use the command `show urgency`. If you use `show` with an options that is shown by default, or `hide` with an option that is hidden by default, this will ne a no-op.

That is useful information for users - would be good to see that in the docs somehow.

It would also be good for the Quick Reference docs to be updated.

## HTML elements

> This PR also slightly simplifies the HTML elements that are postponed to the task. Before, the edit button was inside the span for the backlink, which does not make much sense. The elements for the backlink, the edit button and the urgency score are now simply appended as childs of the existing list item, not nested into each other.

Will the change in nesting of elements:

- affect any existing user CSS customisations in any observable way?
- affect the rendering speed?


## Tests

> Let me know if I shall add tests as well and what would be the best way to do it.

Yes please. There is a test - probably in Query.test.ts - that confirms that all the `hide` commands parse without error.

This should at a minimum have the new `show` command added - but better still would be for it to also test the opposite variants as well.

Cito commented:
Thanks for the quick feedback @claremacrae - all very good points, I'll try to address them this week.

Cito commented:
One thing I currently cannot reproduce is the missing score for "done" tasks. They are visible (stroke through) with both default light and dark themes, tested on Win and Mac. Are you using a different theme?

claremacrae commented:
> One thing I currently cannot reproduce is the missing score for "done" tasks. They are visible (stroke through) with both default light and dark themes, tested on Win and Mac. Are you using a different theme?

Odd. I was using the demo vault, with light theme. On macOS but 2 or 3 year old OS.

Cito commented:
Strange, but I guess it will be fixed when I set an explicit color on that element which I want to do anyway to ensure the colors pass the [conrast checker](https://webaim.org/resources/contrastchecker/).

claremacrae commented:
Oh sorry I’ve just seen that you did update the quick reference page. Thank you. 

I was presuming it would instead go on the same row in the big table as the sort by urgency instruction, to fill in a gap.

Cito commented:
@claremacrae I think I have addressed all your feedback now. Can you have another look?

Regarding rendering speed, I don't see how this should impact performance negatively. Of course, if the scores are shown, they all need to be calculated once, even if the lists are not sorted. But if the urgency scores are not shown, which is the default, there should be no difference. Not sure how I can measure rendering speed easily though.

Regarding CSS, I kept the existing class names and only added one new class `tasks-score` for the urgency score.

claremacrae commented:
> @claremacrae I think I have addressed all your feedback now. Can you have another look?

Thank you for the improvements.

I will review it as soon as I can. Because of the CSS changes, it's going to take longer to review than for single-topic changes - including to figure out how to test the impact of the changes.

claremacrae commented on styles.css:
```diff
@@ -13,11 +13,23 @@
     height: 1em;
     vertical-align: middle;
-    margin-left: 0.3em;
+    margin-left: 0.5em;
```
What was the thinking behind this change please?
(asking as a non-CSS developer)

Cito replied:
I thought the space between the last brace and the pencil was a bit small. It should be at bit larger than the size of a space char that separates other elements to make it stand out as a separate element, particularly as the pencil is actually not part of the task.

claremacrae replied:
> I thought the space between the last brace and the pencil was a bit small. It should be at bit larger than the size of a space char that separates other elements to make it stand out as a separate element, particularly as the pencil is actually not part of the task.

Nice. 

I would have liked that to be mentioned in the machine-generated changelog for the next release. It's only a tiny change, but it's a nice improvement, and someone may care to know about the diffs.

By it being inside this PR, it won't get mentioned in the release notes, and when I merge this PR, I will have to remember to add an extra bullet point mentioning the extra change, when editing the commit message.

I'll change the summer for this commit, for now.

claremacrae commented on src/QueryRenderer.ts:
```diff
@@ -186,15 +186,18 @@ class QueryRenderChild extends MarkdownRenderChild {
+
             if (!this.query.layoutOptions.hideBacklinks) {
-                this.addBacklinks(postInfo, task, shortMode, isFilenameUnique);
+                this.addBacklinks(listItem, task, shortMode, isFilenameUnique);
```
Let's talk about the change from `postInfo` to `listItem` when we pair on urgency customisation next.

I've some questions about the consequences for users - and they will be much easier to talk about in person than via keyboard!

Cito replied:
OK

claremacrae replied:
Summary: Before this PR, the addBacklinks() method added a 'tasks-backlink' to the element passed in to it.

Now it creates its own nested span and styles that. This simplifies the code...

claremacrae commented:
The scores look great now - thank you!

![image](https://user-images.githubusercontent.com/4840096/190878433-effed934-72ff-4443-827d-3bb324df4b5c.png)

![image](https://user-images.githubusercontent.com/4840096/190878438-2c1b849d-a9a7-4b9b-8b66-91da972c5951.png)

claremacrae commented on docs/advanced/urgency.md:
````diff
@@ -120,3 +120,7 @@ urgency = 0.0 + 6.0 + 5.0 + 0.0 = 11.0
 urgency = 0.0 + 6.0 + 0.0 - 3.0 = 3.0
 ```
+
+## How to Display the Urgency score
````
Nice. I think it's worth adding a table-of-contents to this page now, so people can see at a glance that it is possible to display the scores.

The Layout page shows how to do this.

Cito replied:
Enabled the toc now.

claremacrae replied:
What you did was logical, but did not change the rendered page:

![image](https://user-images.githubusercontent.com/4840096/190888607-127fb474-c46e-441c-973a-20748f7a0f23.png)

I've taken the liberty of pushing a change to the required lines - for info, they are:

```text
{: .no_toc }

<details open markdown="block">
  <summary>
    Table of contents
  </summary>
  {: .text-delta }
1. TOC
{:toc}
</details>
```

claremacrae commented on src/Query/Query.ts:
```diff
@@ -139,38 +139,42 @@ export class Query implements IQuery {
         const hideOptionsMatch = line.match(this.hideOptionsRegexp);
         if (hideOptionsMatch !== null) {
-            const option = hideOptionsMatch[1].trim().toLowerCase();
+            const hide = hideOptionsMatch[1] === 'hide';
```
I had to think quite hard about the logic on line 142 and 174. I didn't spot the `!` on 174 the first time I read the code.

I could certainly live with this code remaining as it is, if you wanted to put this PR down.

But I think that the code would be easier to understand and maintain if the new option was called `hideUrgency` and it started with a default value of true.

That way, the naming and logic is the same for all the options, which will reduce the chances of bugs creeping in, in future.

Cito replied:
Yes, I think you're right. The small benefit of having the default value to be always false is not worth it. Changed it now.

claremacrae reviewed (approved):
Just a few minor comments and a question.

It's totally up to you whether you would like to pick up the possible changes I mentioned... 

(I'm happy to do them, if not...)

claremacrae commented on styles.css:
```diff
@@ -13,11 +13,23 @@
 
-.internal-link.internal-link-short-mode {
+/* Urgency score */
+.tasks-score {
```
I've just noticed the difference between the comment and the name `.tasks-score`.

If I were looking to style the urgency, I would look for `urgency` in the name - plus, who knows, maybe there might be some other kind of score for tasks in the future?

So I think one of these is probably a better name for the addition:

- `.tasks-urgency`
- `.tasks-urgency-score`

Also, that would make it more consistent with the name of the instruction to show/hide it.

Cito replied:
I tried to avoid the term "urgency" because of [these thoughts](https://github.com/obsidian-tasks-group/obsidian-tasks/discussions/1157), but I agree that should be a separate discussion which I have started there.

For the time being, we should stick to the term "urgency" to be consistent, so I changed this.

claremacrae replied:
It has been renamed to `.tasks-urgency`

claremacrae commented on resources/sample_vaults/Tasks-Demo/.obsidian/plugins/obsidian-tasks-plugin/styles.css:
```diff
@@ -13,11 +13,23 @@
+    line-height: 1;
+}
+
+a.internal-link.internal-link-short-mode {
```
Ah, I missed this diff several times in previous reading of the changes... It got buried in the diff, because it was adjacent to the added text - I only just noticed it.

```diff
- .internal-link.internal-link-short-mode {
+ a.internal-link.internal-link-short-mode {
```

For info, it's what I would like to talk about when we next chat... (no need to reply now)

claremacrae replied:
That change is going to be reverted... It turns out not to be necessary. (To test, make sure there is no underline on the chain-link icon for backlink in short mode.

claremacrae replied:
Now reverted, in both copies of the styles.css file

claremacrae reviewed (commented):
A couple more thoughts.

claremacrae commented on docs/queries/layout.md:
```diff
@@ -20,12 +20,14 @@ parent: Queries
 
 ## Hiding Elements
```
I wonder whether it is worth updating this heading now it's possible to show elements as well?

Cito replied:
Ok, done

claremacrae commented:
Amazing work - huge thanks @Cito !
